metric_learn/base_metric.py

```diff
@@ -516,26 +536,45 @@ def calibrate_threshold(self, pairs_valid, y_valid, strategy='accuracy',
       f_beta[np.isnan(f_beta)] = 0.
       imax = np.argmax(f_beta)
       # note: we want a positive threshold (distance), so we take - threshold
-      self.threshold_ = - thresholds[imax]
+      if imax == len(thresholds):  # the best is to accept all points
+        # we set the threshold to (minus) [the lowest score - 1]
+        self.threshold_ = - (thresholds[imax] - 1)
+      else:
+        # otherwise, we set the threshold to the mean between the lowest
+        # accepted score and the highest rejected score
+        self.threshold_ = - np.mean(thresholds[imax: imax + 2])
+      # Note: we don't need to deal with rejecting all points (i.e. threshold =
+      # max_scores + 1), since this can never happen to be optimal
+      # (see a more detailed discussion in test_calibrate_threshold_extreme)
       return self
 
     fpr, tpr, thresholds = roc_curve(y_valid,
                                      self.decision_function(pairs_valid),
-                                     pos_label=1)
+                                     pos_label=1, drop_intermediate=False)
```

Comment:
We need `drop_intermediate=False` here for a similar problem than above: it worked fine when we were setting the threshold to an exact value of a point, but if we remove some points the mean between one point and the next can be wrong (because the "next" point is not the actual "next" point: the actual "next" point can be removed)